### PR TITLE
8304442: Allocate VirtualMemoryTracker into Arena

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

This is a suggestion to allocate the VirtualMemoryTracker memory inside of an Arena instead of on the heap. This reduces the number of NativeCallStacks allocated as VMT doesn't go through os::malloc for each linked list node. It also hopefully increases memory locality, as the nodes are in the best case allocated very close to each other.

This PR sees a performance improvement in os::commit_memory and reserve_memory of about 10-25%, at no cost to reporting.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304442](https://bugs.openjdk.org/browse/JDK-8304442): Allocate VirtualMemoryTracker into Arena

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13190/head:pull/13190` \
`$ git checkout pull/13190`

Update a local copy of the PR: \
`$ git checkout pull/13190` \
`$ git pull https://git.openjdk.org/jdk.git pull/13190/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13190`

View PR using the GUI difftool: \
`$ git pr show -t 13190`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13190.diff">https://git.openjdk.org/jdk/pull/13190.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13190#issuecomment-1484981886)

### CODE REVIEW

File: src/hotspot/share/services/virtualMemoryTracker.cpp [hunk 1]
```diff
@@ -107,6 +108,25 @@ static bool try_merge_with(LinkedListNode<CommittedMemoryRegion>* node, LinkedLi
...
+ReservedMemoryRegion::ReservedMemoryRegion(address base, size_t size, const NativeCallStack& stack, MEMFLAGS flag) :
+  VirtualMemoryRegion(base, size),
+  _committed_regions(VirtualMemoryTracker::_backing_arena),
+  _stack(stack), _flag(flag) { }
+
+
+ReservedMemoryRegion::ReservedMemoryRegion(address base, size_t size) :
+  VirtualMemoryRegion(base, size),
+  _committed_regions(VirtualMemoryTracker::_backing_arena),
+  _stack(NativeCallStack::empty_stack()), _flag(mtNone) { }
```
Comment (coleenp) [inline src/hotspot/share/services/virtualMemoryTracker.cpp:120]: You shouldn't record anything with mtNone (not sure if this is what this does).

File: src/hotspot/share/services/virtualMemoryTracker.cpp [hunk 2]
```diff
@@ -107,6 +108,25 @@ static bool try_merge_with(LinkedListNode<CommittedMemoryRegion>* node, LinkedLi
...
+
+// Copy constructor
+ReservedMemoryRegion::ReservedMemoryRegion(const ReservedMemoryRegion& rr) :
+    VirtualMemoryRegion(rr.base(), rr.size()),
+    _committed_regions(VirtualMemoryTracker::_backing_arena) {
+    *this = rr;
+}
```
Comment (coleenp) [inline src/hotspot/share/services/virtualMemoryTracker.cpp:127]: (preexisting) since you have a copy constructor, should you have a deleted operator=?

**Review by coleenp — COMMENTED** [review]
These sorted regions are never deleted?

File: src/hotspot/share/services/virtualMemoryTracker.cpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/share/services/virtualMemoryTracker.cpp:127]: Yeah, I also thought it was weird that Rule-of-3 isn't obeyed here. Just at a glance, seems like the default destructor does the right thing (calling destructor for each member).

**Comment by tstuefe** [comment]
> > These sorted regions are never deleted?
> 
> They are, via `VMT::remove_released_region`.

I don't understand then; would this not leak memory if we randomly remove regions? Since with an Arena, you only can effectively reclaim the topmost allocation?

Easy check: call os::reserve_memory/commit_memory a million times with random - and disjunct - address ranges. If RSS stays stable, we are good.

**Comment by jdksjolen** [comment]
Hi!

Sorry for the late response, I've been busy with a lot of other things. There's no confusion on Thomas's or Coleen's part, it's me who was confused and forgot the specific semantics of `Arena` unfortunately :). This PR does indeed leak memory, because of Arena's inability to free part-of-a-chunk. I hope I didn't waste your time too much on this! I'll close this, as it's clearly not a sound PR.

**Pull request closed without merging**